compiler/rustc_metadata/src/locator.rs

```diff
@@ -783,14 +818,69 @@ fn get_metadata_section<'p>(
     filename: &'p Path,
     loader: &dyn MetadataLoader,
     cfg_version: &'static str,
+    dyn_crate: Option<Symbol>,
 ) -> Result<MetadataBlob, MetadataError<'p>> {
     if !filename.exists() {
         return Err(MetadataError::NotPresent(filename));
     }
+
+    if dyn_crate.is_some() && flavor != CrateFlavor::SDylib {
+        return Err(MetadataError::LoadFailure(
```

Comment:
This can be an assert, not a user error, we control both `dyn_crate` and `flavor` when calling `get_metadata_section`.